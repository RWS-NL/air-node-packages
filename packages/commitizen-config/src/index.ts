import { configLoader } from 'commitizen';
import commitTypes from './commit-types';
import engine from './engine';
import commitlintLoad from '@commitlint/load';

const config = configLoader.load();
const options = {
  types: commitTypes.types,
  defaultType: process.env.CZ_TYPE || config.defaultType,
  defaultScope: process.env.CZ_SCOPE || config.defaultScope,
  defaultSubject: process.env.CZ_SUBJECT || config.defaultSubject,
  defaultBody: process.env.CZ_BODY || config.defaultBody,
  defaultIssues: process.env.CZ_ISSUES || config.defaultIssues,
  disableScopeLowerCase: process.env.DISABLE_SCOPE_LOWERCASE || config.disableScopeLowerCase,
  maxHeaderWidth:
    (process.env.CZ_MAX_HEADER_WIDTH && parseInt(process.env.CZ_MAX_HEADER_WIDTH)) || config.maxHeaderWidth || 100,
  maxLineWidth: (process.env.CZ_MAX_LINE_WIDTH && parseInt(process.env.CZ_MAX_LINE_WIDTH)) || config.maxLineWidth || 100
};

(function (options) {
  try {
    commitlintLoad().then(function (clConfig: { rules: { [x: string]: any } }) {
      if (clConfig.rules) {
        const maxHeaderLengthRule = clConfig.rules['header-max-length'];
        if (
          typeof maxHeaderLengthRule === 'object' &&
          maxHeaderLengthRule.length >= 3 &&
          !process.env.CZ_MAX_HEADER_WIDTH &&
          !config.maxHeaderWidth
        ) {
          options.maxHeaderWidth = maxHeaderLengthRule[2];
        }
      }
    });
  } catch (err: any) {
    throw new Error(err);
  }
})(options);

export default engine(options);
