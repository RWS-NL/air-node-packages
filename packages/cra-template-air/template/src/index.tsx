import { isDevEnv } from '@rws-air/utils';
import App from 'App';
import 'config/i18n';
import 'core-js';
import React from 'react';
import ReactDOM from 'react-dom';
import 'styles/styles.scss';
import 'whatwg-fetch';
import { HotNodeModule } from '{{APP_NAME_REDUX}}';

const target = document.querySelector('#root');
ReactDOM.render(<App />, target);

if ('serviceWorker' in navigator) {
  navigator.serviceWorker.ready.then((registration) => {
    registration.unregister();
  });
}

if (isDevEnv) {
  if ((module as HotNodeModule).hot) {
    (module as HotNodeModule).hot.accept('./App.tsx', () => {
      // eslint-disable-next-line
      const NextApp = require('./App').default;
      ReactDOM.render(<NextApp />, target);
    });
  }
}
