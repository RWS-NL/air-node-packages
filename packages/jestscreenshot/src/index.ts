import moment from 'moment';
import mkdirp from 'mkdirp';
import fetch from 'node-fetch';
import Form from 'form-data';
import path from 'path';
import { Page } from 'puppeteer';

export class JestScreenshotError extends Error {
  /**
   * Constructs a new JestScreenshotError
   * @param message Message to append to standard message for JestScreenshotError
   */
  public constructor(message: string) {
    super();

    this.message = `Failed to create JestScreenshot Reporter.\n${message}`;
  }
}

export interface JestScreenshotOptions {
  /** The Puppeteer page to take a screenshot from */
  page: Page;
  /** The directory to create a "screenshots" folder in */
  dirName: string;
  /**
   * @description An optional name of the script that is currently being ran
   * @default jest-test
   */
  testName?: string;
  /**
   * Optionally upload screenshots to slack after making them
   * Requires you pass a token to the slackToken option, or set the SLACK_WEBTOKEN environment variable
   * @default false
   */
  slackUpload?: boolean;
  /**
   * Token to use when uploading to Slack
   * Only used when slackUpload is set to true
   * Recommended to pass this as SLACK_WEBTOKEN environment variable instead!
   * @default ''
   */
  slackToken?: string;

  /**
   * Channels to send the Slack upload to
   * Should be an array of Slack channel IDs
   * Only used when slackUpload is set to true
   * @example
   * ```typescript
   * {
   *   slackChannels: ['C1234567890', 'C2345678901', 'C3456789012']
   * }
   * ```
   */
  slackChannels?: string[];
}

/** Represents all data returned by Slack when uploading a file */
export interface SlackFile {
  id: string;
  created: number;
  timestamp: number;
  name: string;
  title: string;
  mimetype: string;
  filetype: string;
  pretty_type: string;
  user: string;
  editable: boolean;
  size: number;
  mode: string;
  is_external: boolean;
  external_type: string;
  is_public: boolean;
  public_url_shared: boolean;
  display_as_bot: boolean;
  username: string;
  url_private: string;
  url_private_download: string;
  thumb_64: string;
  thumb_80: string;
  thumb_360: string;
  thumb_360_w: number;
  thumb_360_h: number;
  thumb_480: string;
  thumb_480_w: number;
  thumb_480_h: number;
  thumb_160: string;
  image_exif_rotation: number;
  original_w: number;
  original_h: number;
  permalink: string;
  permalink_public: string;
  comments_count: number;
  is_starred: boolean;
  shares: {
    private: {
      [key: string]: {
        reply_users: string[];
        reply_users_count: number;
        reply_count: number;
        ts: string;
      };
    };
  };
  channels: string[];
  groups: string[];
  ims: string[];
  has_rich_preview: boolean;
}

/** Represents the response by Slack when uploading a file */
export interface SlackResponse {
  ok: boolean;
  file: SlackFile;
}

/**
 * The main class of JestScreenshot that should be initialized with config
 *
 * @remarks Jest reporter plugin to take Puppeteer screenshots on failing tests
 *
 * @param page The Puppeteer page object to screenshot
 * @param dirName The directory to create a "screenshots" folder in
 * @param testName An optional name of the script that is currently being ran
 * @param slackUpload Optionally upload screenshots to slack after making them.
 *
 * Requires you pass a token to the slackToken option, or set the SLACK_WEBTOKEN environment variable
 * @param slackToken Token to use when uploading to slack. Required when you pass slackUpload=true
 *
 * Optionally you can also pass this through the environment variable "SLACK_WEBTOKEN".
 *
 * This environment variable will take priority over passing it as option
 *
 * @param slackChannels Channels to send the Slack upload to
 *
 * Should be an array of Slack channel IDs
 *
 * Only used when slackUpload is set to true
 */
export class JestScreenshot {
  private page: Page;
  private dirName: string;
  private testName: string;
  private shouldUploadToSlack = false;
  private slackToken: string | undefined = '';
  private slackChannels: string[] | undefined = [];

  /**
   * Constructs a new JestScreenshot
   * @param options The options to pass to the instance of JestScreenshot
   */
  public constructor(options: JestScreenshotOptions) {
    if (!this.objectHasProperty(options, 'page')) {
      throw new JestScreenshotError('You should pass page to screenshot to options (page)!!');
    }
    if (!this.objectHasProperty(options, 'dirName')) {
      throw new JestScreenshotError('You should pass the name of the directory to save images to the options (dirName)!!');
    }

    this.page = options.page;
    this.dirName = `${options.dirName}/screenshots`;
    this.testName = options.testName && 'testName' in options ? options.testName : 'jest-test';
    this.shouldUploadToSlack = options.slackUpload && 'slackUpload' in options ? options.slackUpload : false;

    if (this.shouldUploadToSlack) {
      if (process.env.SLACK_WEBTOKEN) {
        this.slackToken = process.env.SLACK_WEBTOKEN;
      } else {
        this.slackToken = options.slackToken;
      }

      if (!this.slackToken) throw new JestScreenshotError('When you want to upload to slack you should either provide the slackToken option or an environment variable called "SLACK_WEBTOKEN"');

      if (!this.objectHasProperty(options, 'slackChannels')) {
        throw new JestScreenshotError('When you want to upload to slack you provide an array of channels to post the screenshot in through the "slackChannels" options');
      }

      this.slackChannels = options.slackChannels;

      if (!this.slackChannels || !this.slackChannels.length) {
        throw new JestScreenshotError('Failed to set the Slack channels, please verify your slackChannels option');
      }
    }
  }

  /**
   * Sets up the JestScreenshot reporter
   *
   * @returns JestScreenshot reporter will have been initialiazed as a side effect for this test suite
   */
  public async setup(): Promise<void> {
    try {
      mkdirp.sync(this.dirName);
    } catch {
      throw new Error('Failed to create screenshots directory, maybe check your permissions?');
    }

    await this.takeScreenshot();
  }

  /**
   * Takes a screenshot of the current page
   * @returns Either returns the response from Slack or the screenshot as a Buffer
   */
  public async takeScreenshot(): Promise<SlackResponse | Buffer> {
    const fileName = `${this.testName}-${moment().format('YYYY-MM-DD[_]HH.mm.ss')}.png`;
    const filePath = path.join(this.dirName, fileName);

    const screenshot = await this.page.screenshot({ path: filePath });

    if (this.shouldUploadToSlack) return this.uploadToSlack(screenshot, fileName.slice(0, -4));

    return screenshot;
  }

  /**
   * Uploads screenshots to Slack using the provided token
   *
   * @param screenshot base64 representation of the screenshot to upload
   * @returns {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise | Promise} of the SlackResponse
   */
  public async uploadToSlack(screenshot: Buffer, fileName: string): Promise<SlackResponse> {
    const apiUrl = 'https://slack.com/api';
    const apiMethod = `${apiUrl}/files.upload`;
    const form = new Form();
    form.append('token', this.slackToken);
    form.append('channels', this.slackChannels?.join(',') || [].join(','));
    form.append('file', screenshot, {
      contentType: 'image/png',
      filename: fileName,
    });
    form.append('filename', fileName);
    form.append('title', fileName);
    form.append('filetype', 'png');

    try {
      const request = await fetch(apiMethod, {
        method: 'POST',
        body: form,
      });

      const data = await request.json() as SlackResponse;

      return data;
    } catch (err) {
      throw new Error(err);
    }
  }

  /**
   * Validates if an object has a given property
   * @param obj The object to validate
   * @param prop The property to try to find
   * @returns Whether the object has this property or not
   */
  public objectHasProperty<O extends object, K extends keyof O>(obj: O, prop: K): boolean {
    return obj && prop in obj;
  }
}

export default JestScreenshot;
module.exports = JestScreenshot;