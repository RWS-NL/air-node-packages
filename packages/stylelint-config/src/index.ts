export const config = {
  defaultSeverity: 'error',
  extends: [
    'stylelint-config-standard',
    'stylelint-config-recommended',
    'stylelint-config-sass-guidelines'
  ],
  plugins: [
    'stylelint-scss'
  ],
  rules: {
    'color-hex-case': 'upper',
    'color-hex-length': 'long',
    'color-named': 'never',
    'declaration-block-no-redundant-longhand-properties': true,
    'declaration-block-no-shorthand-property-overrides': true,
    'declaration-empty-line-before': 'never',
    'font-family-name-quotes': 'always-where-recommended',
    'function-calc-no-invalid': true,
    'max-nesting-depth': [
      3,
      {ignoreAtRules: [ 'each', 'media', 'supports', 'include' ]}
    ],
    'no-invalid-double-slash-comments': null,
    'number-max-precision': 2,
    'property-no-unknown': true,
    'scss/at-else-closing-brace-newline-after': 'always-last-in-chain',
    'scss/at-else-closing-brace-space-after': 'always-intermediate',
    'scss/at-else-empty-line-before': 'never',
    'scss/at-else-if-parentheses-space-before': 'always',
    'scss/at-if-closing-brace-newline-after': 'always-last-in-chain',
    'scss/at-if-closing-brace-space-after': 'always-intermediate',
    'scss/at-mixin-argumentless-call-parentheses': 'never',
    'scss/at-mixin-parentheses-space-before': 'always',
    'scss/at-mixin-pattern': "^[a-z](?:[A-Z0-9]*[a-z][a-z0-9]*[A-Z]|[a-z0-9]*[A-Z][A-Z0-9]*[a-z])?[A-Za-z0-9]*$",
    'scss/dollar-variable-no-missing-interpolation': true,
    'scss/dollar-variable-pattern': "^[a-z](?:[A-Z0-9]*[a-z][a-z0-9]*[A-Z]|[a-z0-9]*[A-Z][A-Z0-9]*[a-z])?[A-Za-z0-9]*$",
    'scss/double-slash-comment-whitespace-inside': 'always',
    'scss/no-duplicate-dollar-variables': true,
    'scss/percent-placeholder-pattern': "^[a-z](?:[A-Z0-9]*[a-z][a-z0-9]*[A-Z]|[a-z0-9]*[A-Z][A-Z0-9]*[a-z])?[A-Za-z0-9]*$",
    'selector-class-pattern': "^[a-z](?:[A-Z0-9]*[a-z][a-z0-9]*[A-Z]|[a-z0-9]*[A-Z][A-Z0-9]*[a-z])?[A-Za-z0-9]*$",
    'value-keyword-case': 'lower',
    'value-list-comma-newline-after': 'never-multi-line',
    'value-list-comma-newline-before': 'never-multi-line',
  },
};

module.exports = config;
exports.default = config;
export default config;