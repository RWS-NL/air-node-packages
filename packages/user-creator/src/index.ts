#!/usr/bin/env node
import { readYaml } from '@favware/yamlreader';
import fetch from 'node-fetch';
import path from 'path';
import yargs from 'yargs';

export interface User {
  /** The e-mail for this user */
  email: string;
  /** The firstname for this user */
  firstname: string;
  /** The optional middle name for this user */
  middlename?: string;
  /** The lastname for this user */
  lastname: string;
  /** Any roles to assign to the user */
  roles: string[];
}

interface ResponseData extends User {
  /** ID of the user in the database */
  id: string;
  /** Privileges the user has */
  privileges: string[];
}

export class UsermanagementAdder {
  private file: string;
  private username: string;
  private password: string;
  private url: string;

  /**
   * Constructs a new {@link UsermanagementAdder}
   * @param file The YAML file of userdata to read
   * @param user The admin user name to create users with
   * @param pass The password for the admin user
   * @param url The URL of the usermanagement backend
   */
  public constructor(file: string, user: string, pass: string, url: string) {
    this.file = file;
    this.username = user;
    this.password = pass;
    this.url = url;
  }

  /**
   * Runs the {@link UsermanagementAdder}
   */
  public async run() {
    const users: User[] = readYaml(this.file);
    for (const user of users) await this.addUserToAPI(user);
  }

  private async addUserToAPI(user: User) {
    try {
      const response = await fetch(`${this.url}/api/v1/users`, {
        body: JSON.stringify({
          email: user.email,
          firstName: user.firstname,
          lastName: user.lastname,
          middleName: user.middlename ? user.middlename : ''
        }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Basic ${Buffer.from(`${this.username}:${this.password}`).toString('base64')}`
        }
      });
      const data: ResponseData = await response.json() as ResponseData;

      console.log(`Added user with email ${user.email}, they got ID ${data.id}`);

      for (const role of user.roles) await this.addRoleToUser(data.id, role);
    } catch (err) {
      console.error(err);
    }
  }

  private async addRoleToUser(userId: string, role: string) {
    console.log(`Adding ${role} to user with ID ${userId}`);

    try {
      await fetch(`${this.url}/api/v1/users/${userId}/roles`, {
        body: JSON.stringify({ name: role }),
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Basic ${Buffer.from(`${this.username}:${this.password}`).toString('base64')}`
        }
      });

      console.log(`Added role ${role} to user with id ${userId}`);
      console.log('');
    } catch (err) {
      console.error(`an error occurred adding ${role} to user with ID ${userId}. Stacktrace:`);
      console.debug(err);
    }
  }
}

/**
 * Executes the usermanagement usercreator
 * Uses yargs to parse arguments
 */
export const exec = () => {
  const argv = yargs
    .usage(
      [
        'Usermanagement user adding script',
        '',
        '\tUsage:',
        '\t\tusercreator -f /path/to/yaml/file -u <username> -p <password> --url <url>',
        '\t\tusercreator --help'
      ].join('')
    )
    .example('', 'usercreator -f users.yaml -u root -p root --url http://localhost:9003')
    .option('file', {
      alias: 'f',
      describe: 'The YAML file to use as users list',
      type: 'string',
      demand: true
    })
    .option('username', {
      alias: 'u',
      describe: 'The username to use as "root" user when uploading',
      type: 'string',
      demand: true
    })
    .option('password', {
      alias: 'p',
      describe: 'The password to use as "root" user when uploading',
      type: 'string',
      demand: true
    })
    .option('url', {
      alias: 's',
      describe: 'The url to upload to',
      type: 'string',
      demand: true
    }).argv;

  if (argv instanceof Promise) {
    console.error('argv is of type Promise which is not supported yet');
  } else {
    if (argv.url.slice(-1) === '/') argv.url = argv.url.slice(0, -1);

    try {
      new UsermanagementAdder(path.resolve(argv.file), argv.username, argv.password, argv.url).run();
    } catch (err) {
      console.error(err);
    }
  }
};

exec();

export default exec;
