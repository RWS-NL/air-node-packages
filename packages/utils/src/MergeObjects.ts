import { isObject } from './IsObject';

export type KeyedObject = Record<PropertyKey, unknown>;

/**
 * Merges two objects
 * @param objTarget The object to be merged
 * @param objSource The object to merge
 */
export function mergeObjects<A, B>(objTarget: A, objSource: B): A & B {
  for (const [key, value] of Object.entries(objSource) as [keyof B, unknown][]) {
    const targetValue = (objTarget as KeyedObject)[key];
    if (isObject(value)) {
      Reflect.set(
        objTarget as KeyedObject,
        key,
        isObject(targetValue) ? mergeObjects(targetValue as KeyedObject, value as KeyedObject) : value
      );
    } else if (!isObject(targetValue)) {
      Reflect.set(objTarget as KeyedObject, key, value);
    }
  }
  return objTarget as A & B;
}
