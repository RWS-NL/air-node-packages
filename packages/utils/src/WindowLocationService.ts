export class WindowLocationService {
  public static get origin(): string {
    return this.getLocation().origin;
  }

  public static get path(): string {
    return this.getLocation().pathname;
  }

  public static get hash(): string {
    return this.getLocation().hash;
  }

  public static get search(): string {
    return this.getLocation().search;
  }

  public static get host(): string {
    return this.getLocation().host;
  }

  public static get href(): string {
    return this.getLocation().href;
  }

  public static redirectTo(url: string): void {
    this.getLocation().assign(url);
  }

  private static getLocation(): Location {
    return window.location;
  }
}
