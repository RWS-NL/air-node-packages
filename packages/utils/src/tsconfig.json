{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "../dist",
    "tsBuildInfoFile": "../dist/.tsbuildinfo",
    "skipLibCheck": true
  },
  "include": ["./"]
}
