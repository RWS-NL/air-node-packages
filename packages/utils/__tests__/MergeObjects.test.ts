import { mergeObjects } from '../src/MergeObjects';

describe('Merge Objects', () => {
  test('mergeObjects(basic)', () => {
    const source = { a: 0, b: 1 };
    const target = {};
    expect(mergeObjects(target, source)).toStrictEqual({ a: 0, b: 1 });
    expect(mergeObjects(target, source)).toMatchSnapshot();
  });

  test('mergeObjects(mutation)', () => {
    const source = { a: 0, b: 1 };
    const target = {};
    mergeObjects(target, source);

    expect(source).toStrictEqual({ a: 0, b: 1 });
    expect(target).toStrictEqual({ a: 0, b: 1 });
    expect(source).toMatchSnapshot();
    expect(target).toMatchSnapshot();
  });

  test('mergeObjects(clone)', () => {
    const source = { a: 0, b: 1 };
    const target = {};
    expect(mergeObjects(target, source)).toStrictEqual({ a: 0, b: 1 });
    expect(mergeObjects(target, source)).toMatchSnapshot();
  });

  test('mergeObjects(partial)', () => {
    const source = { a: 0, b: 1 };
    const target = { a: 2 };
    expect(mergeObjects(target, source)).toStrictEqual({ a: 0, b: 1 });
    expect(mergeObjects(target, source)).toMatchSnapshot();
  });

  test('mergeObjects(extended)', () => {
    const source = { a: 0, b: 1 };
    const target = { a: 2, i: 2 };
    expect(mergeObjects(target, source)).toStrictEqual({ a: 0, i: 2, b: 1 });
    expect(mergeObjects(target, source)).toMatchSnapshot();
  });

  test('Function Snapshot', () => {
    expect(mergeObjects.toString()).toMatchSnapshot();
  });
});
