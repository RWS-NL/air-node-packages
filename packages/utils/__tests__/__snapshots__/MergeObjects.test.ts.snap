// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge Objects Function Snapshot 1`] = `
"function mergeObjects(objTarget, objSource) {
    for (const [key, value] of Object.entries(objSource)) {
        const targetValue = objTarget[key];
        if ((0, IsObject_1.isObject)(value)) {
            Reflect.set(objTarget, key, (0, IsObject_1.isObject)(targetValue) ? mergeObjects(targetValue, value) : value);
        }
        else if (!(0, IsObject_1.isObject)(targetValue)) {
            Reflect.set(objTarget, key, value);
        }
    }
    return objTarget;
}"
`;

exports[`Merge Objects mergeObjects(basic) 1`] = `
Object {
  "a": 0,
  "b": 1,
}
`;

exports[`Merge Objects mergeObjects(clone) 1`] = `
Object {
  "a": 0,
  "b": 1,
}
`;

exports[`Merge Objects mergeObjects(extended) 1`] = `
Object {
  "a": 0,
  "b": 1,
  "i": 2,
}
`;

exports[`Merge Objects mergeObjects(mutation) 1`] = `
Object {
  "a": 0,
  "b": 1,
}
`;

exports[`Merge Objects mergeObjects(mutation) 2`] = `
Object {
  "a": 0,
  "b": 1,
}
`;

exports[`Merge Objects mergeObjects(partial) 1`] = `
Object {
  "a": 0,
  "b": 1,
}
`;
